### PR TITLE
Python3 pickled machine learning (scikit-learn) models

### PR DESCRIPTION
**Pull request by anuprulez**
This pull request writes the trained machine learning models as pickled models. The version of python is migrated from 2.7 to 3.6.

### CODE REVIEW

File: tools/sklearn/utils.py [hunk 1]
```diff
@@ -60,17 +59,16 @@ def find_class(self, module, name):
...
 
-    @classmethod
-    def load(self, file):
-        obj = pickle.Unpickler(file)
-        obj.find_global = self.find_class
-        return obj.load()
+
+def load_model(file):
+    return SafePickler(file).load()
+
```
Comment (qiagu) [inline tools/sklearn/utils.py]: The old `SafePickler.load` was removed. The new `load_model` seems to be the same as `pickle.load`. The  `SafePickler.find_class` which contains the whitelist checker is not applied at all. Please double check this.

File: tools/sklearn/utils.py [hunk 2]
```diff
@@ -43,15 +42,15 @@ def find_class(self, module, name):
 
         if re.match(r'^[a-zA-Z_][a-zA-Z0-9_]*$', name):
             fullname = module + '.' + name
-            if  (fullname in good_names)\
+            if (fullname in good_names)\
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/utils.py]: 'copyreg' and 'builtins' ?

**Comment by qiagu** [comment]
Great work. @anuprulez  Thanks.

File: tools/sklearn/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/utils.py]: It seems these two are not necessary in python3 pickle. I was wrong.

**Comment by anuprulez** [comment]
@qiagu Merged the PR. Thanks!

**Comment by bgruening** [comment]
Thanks all!

**Pull request merged**